File: examples/r_package/tests/testthat/test_unit.R
# Unit testing examples for import_patient_data


test_that("small CSV with correct columns imports successfully", {
  expected_cols <- c(
    "PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME",
    "SERVICE_DATE", "SERVICE_TIME"
  )

  # Create temporary CSV file
  df_in <- tibble::tibble(
    PATIENT_ID   = "p1",
    ARRIVAL_DATE = lubridate::ymd("2024-01-01"),
    ARRIVAL_TIME = hms::as_hms("08:00:00"),
    SERVICE_DATE = lubridate::ymd("2024-01-01"),
    SERVICE_TIME = hms::as_hms("09:00:00")
  )
  csv_path <- tempfile(fileext = ".csv")
  readr::write_csv(df_in, csv_path)

  # Run function and check it looks correct
  result <- import_patient_data(csv_path)
  expect_s3_class(result, "data.frame")
  expect_identical(names(result), expected_cols)
  expect_equal(as.data.frame(result), as.data.frame(df_in))
})


patrick::with_parameters_test_that(
  "incorrect columns cause import_patient_data() to fail",
  {
    # Create dataframe with incorrect columns
    df <- as.data.frame(as.list(seq_along(cols)))
    names(df) <- cols
    # Save as temporary CSV and run function, expecting an error
    csv_path <- tempfile(fileext = ".csv")
    readr::write_csv(df, csv_path)
    expect_error(import_patient_data(csv_path))
  },
  patrick::cases(
    # Example 1: Missing columns
    list(
      cols = c("PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME", "SERVICE_DATE")
    ),
    # Example 2: Extra columns
    list(
      cols = c(
        "PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME", "SERVICE_DATE",
        "SERVICE_TIME", "EXTRA"
      )
    ),
    # Example 3: Right columns, wrong order
    list(
      cols = c(
        "ARRIVAL_DATE", "PATIENT_ID", "ARRIVAL_TIME",
        "SERVICE_DATE", "SERVICE_TIME"
      )
    )
  )
)


test_that("empty CSV with correct columns should succeed", {
  # Empty CSV with correct columns should succeed.

  expected_cols <- c(
    "PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME",
    "SERVICE_DATE", "SERVICE_TIME"
  )

  # Create empty CSV with correct header
  df_in <- tibble::tibble(
    PATIENT_ID   = character(),
    ARRIVAL_DATE = character(),
    ARRIVAL_TIME = character(),
    SERVICE_DATE = character(),
    SERVICE_TIME = character()
  )
  csv_path <- tempfile(fileext = ".csv")
  readr::write_csv(df_in, csv_path)

  # Should succeed and return empty data frame
  result <- import_patient_data(csv_path)
  expect_identical(nrow(result), 0L)
  expect_identical(names(result), expected_cols)
})

File: examples/python_package/tests/test_unit.py
"""
Unit testing examples for import_patient_data.
"""

import pandas as pd
import pytest

from waitingtimes.patient_analysis import import_patient_data


def test_import_success(monkeypatch):
    """Small CSV with correct columns should work."""

    # Create sample patient data
    expected_cols = [
        "PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME",
        "SERVICE_DATE", "SERVICE_TIME",
    ]
    testdata = pd.DataFrame(
        [["p1", "2024-01-01", "08:00", "2024-01-01", "09:00"]],
        columns=expected_cols,
    )

    # Call function (with mocking for pd.read_csv())
    def mock_read_csv(*args, **kwargs):
        return testdata
    monkeypatch.setattr(pd, "read_csv", mock_read_csv)
    result = import_patient_data("path.csv")

    # Check the result looks correct
    assert isinstance(result, pd.DataFrame)
    assert list(result.columns) == expected_cols
    pd.testing.assert_frame_equal(result, testdata)


@pytest.mark.parametrize(
    "columns",
    [
        # Example 1: Missing columns
        [
            "PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME", "SERVICE_DATE"
        ],
        # Example 2: Extra columns
        [
            "PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME",
            "SERVICE_DATE", "SERVICE_TIME", "EXTRA",
        ],
        # Example 3: Right columns, wrong order
        [
            "ARRIVAL_DATE", "PATIENT_ID", "ARRIVAL_TIME",
            "SERVICE_DATE", "SERVICE_TIME",
        ],
    ],
)
def test_import_errors(monkeypatch, columns):
    """Incorrect columns should trigger ValueError."""

    # Create sample patient data
    testdata = pd.DataFrame([range(len(columns))], columns=columns)

    # Call function (with mocking for pd.read_csv()), should raise an error
    def mock_read_csv(*args, **kwargs):
        return testdata
    monkeypatch.setattr(pd, "read_csv", mock_read_csv)
    with pytest.raises(ValueError):
        import_patient_data("path.csv")


def test_import_empty_csv(monkeypatch):
    """Empty CSV with correct columns should succeed."""

    # Create empty CSV with correct header
    expected_cols = [
        "PATIENT_ID", "ARRIVAL_DATE", "ARRIVAL_TIME",
        "SERVICE_DATE", "SERVICE_TIME",
    ]
    testdata = pd.DataFrame(columns=expected_cols)

    # Call function (with mocking for pd.read_csv())
    def mock_read_csv(*args, **kwargs):
        return testdata
    monkeypatch.setattr(pd, "read_csv", mock_read_csv)
    result = import_patient_data("path.csv")

    # Should succeed and return an empty dataframe
    assert len(result) == 0
    assert list(result.columns) == expected_cols


def test_import_path_types(tmp_path):
    """str and Path inputs should behave identically."""
    # Create sample patient data
    expected_cols = [
        "PATIENT_ID",
        "ARRIVAL_DATE", "ARRIVAL_TIME",
        "SERVICE_DATE", "SERVICE_TIME",
    ]
    df_in = pd.DataFrame(
        [["p1", "2024-01-01", "08:00", "2024-01-01", "09:00"]],
        columns=expected_cols,
    )

    # Create temporary file (not mocking, as this is about checking
    # pd.read_csv is working as expected)
    csv_path = tmp_path / "patients.csv"
    df_in.to_csv(csv_path, index=False)

    # Run function with str or Path inputs
    df_str = import_patient_data(str(csv_path))
    df_path = import_patient_data(csv_path)

    # Check that results are the same
    pd.testing.assert_frame_equal(df_str, df_path)
